File: packages/uns-py/src/uns_kit/service_bundle.py
from __future__ import annotations

from dataclasses import dataclass, field
import json
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple


@dataclass
class BundleMetadata:
    name: str
    display_name: Optional[str] = None
    service_type: Optional[str] = None
    summary: Optional[str] = None
    description: Optional[str] = None
    owner: Optional[str] = None
    tags: List[str] = field(default_factory=list)


@dataclass
class BundleScaffold:
    stack: str
    template: str
    features: List[str] = field(default_factory=list)


@dataclass
class BundleRepository:
    provider: Optional[str] = None
    organization: Optional[str] = None
    project: Optional[str] = None
    repository: Optional[str] = None
    default_branch: Optional[str] = None


@dataclass
class BundleDomain:
    inputs: List[Any] = field(default_factory=list)
    outputs: List[Any] = field(default_factory=list)


@dataclass
class BundleServiceSpecDocs:
    goals: List[str] = field(default_factory=list)
    non_goals: List[str] = field(default_factory=list)
    acceptance_criteria: List[str] = field(default_factory=list)


@dataclass
class BundleAgentsDocs:
    project_context: List[str] = field(default_factory=list)
    guardrails: List[str] = field(default_factory=list)
    first_tasks: List[str] = field(default_factory=list)
    verification: List[str] = field(default_factory=list)


@dataclass
class BundleDocs:
    service_spec: BundleServiceSpecDocs = field(default_factory=BundleServiceSpecDocs)
    agents: BundleAgentsDocs = field(default_factory=BundleAgentsDocs)


@dataclass
class ServiceBundle:
    schema_version: int
    kind: str
    metadata: BundleMetadata
    scaffold: BundleScaffold
    repository: Optional[BundleRepository] = None
    domain: Optional[BundleDomain] = None
    docs: BundleDocs = field(default_factory=BundleDocs)


def load_service_bundle(
    bundle_path: Path,
    *,
    expected_stack: str,
    cli_name: str,
    counterpart_cli_name: str,
) -> Tuple[ServiceBundle, str]:
    try:
        raw = bundle_path.read_text()
    except OSError as exc:
        raise ValueError(f"Failed to read bundle JSON at {bundle_path}: {exc}") from exc

    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise ValueError(f"Failed to parse bundle JSON at {bundle_path}: {exc}") from exc

    if not isinstance(parsed, dict):
        raise ValueError("bundle must be a JSON object.")

    bundle = _validate_service_bundle(
        parsed,
        expected_stack=expected_stack,
        cli_name=cli_name,
        counterpart_cli_name=counterpart_cli_name,
    )
    return bundle, raw


def generate_service_spec_markdown(bundle: ServiceBundle) -> str:
    lines = [
        "# SERVICE_SPEC",
        "",
        "> Generated from `service.bundle.json`. Update the bundle as the source of truth.",
        "",
        "## Service Identity",
        *_render_key_value_list(
            [
                ("Name", bundle.metadata.name),
                ("Display Name", bundle.metadata.display_name),
                ("Service Type", bundle.metadata.service_type),
                ("Owner", bundle.metadata.owner),
                ("Tags", ", ".join(bundle.metadata.tags) if bundle.metadata.tags else "None"),
            ]
        ),
        "",
        "## Summary",
        "",
        bundle.metadata.summary or "Not provided.",
        "",
        "## Description",
        "",
        bundle.metadata.description or "Not provided.",
        "",
        "## Scaffold",
        *_render_key_value_list(
            [
                ("Stack", bundle.scaffold.stack),
                ("Template", bundle.scaffold.template),
                ("Features", ", ".join(bundle.scaffold.features) if bundle.scaffold.features else "None"),
            ]
        ),
    ]

    if bundle.repository:
        lines.extend(
            [
                "",
                "## Repository",
                *_render_key_value_list(
                    [
                        ("Provider", bundle.repository.provider),
                        ("Organization", bundle.repository.organization),
                        ("Project", bundle.repository.project),
                        ("Repository", bundle.repository.repository),
                        ("Default Branch", bundle.repository.default_branch),
                    ]
                ),
            ]
        )

    if bundle.domain:
        lines.extend(
            [
                "",
                "## Domain Inputs",
                *_render_unknown_list(bundle.domain.inputs),
                "",
                "## Domain Outputs",
                *_render_unknown_list(bundle.domain.outputs),
            ]
        )

    lines.extend(
        [
            "",
            "## Goals",
            *_render_string_list(bundle.docs.service_spec.goals),
            "",
            "## Non-Goals",
            *_render_string_list(bundle.docs.service_spec.non_goals),
            "",
            "## Acceptance Criteria",
            *_render_string_list(bundle.docs.service_spec.acceptance_criteria),
            "",
        ]
    )
    return "\n".join(lines)


def generate_agents_markdown(bundle: ServiceBundle) -> str:
    lines = [
        "# AGENTS",
        "",
        "> This repository was bootstrapped from `service.bundle.json`. Regenerate derived docs from the bundle instead of treating this file as the source of truth.",
        "",
        "## Service",
        *_render_key_value_list(
            [
                ("Name", bundle.metadata.name),
                ("Display Name", bundle.metadata.display_name),
                ("Stack", bundle.scaffold.stack),
                ("Template", bundle.scaffold.template),
            ]
        ),
        "",
        "## Project Context",
        *_render_string_list(bundle.docs.agents.project_context),
        "",
        "## Guardrails",
        *_render_string_list(bundle.docs.agents.guardrails),
        "",
        "## First Tasks",
        *_render_string_list(bundle.docs.agents.first_tasks),
        "",
        "## Verification",
        *_render_string_list(bundle.docs.agents.verification),
        "",
    ]
    return "\n".join(lines)


def _validate_service_bundle(
    parsed: Dict[str, Any],
    *,
    expected_stack: str,
    cli_name: str,
    counterpart_cli_name: str,
) -> ServiceBundle:
    schema_version = parsed.get("schemaVersion")
    if schema_version != 1:
        raise ValueError(f"service bundle schemaVersion must be 1. Received {schema_version!r}.")

    kind = parsed.get("kind")
    if kind != "uns-service-bundle":
        raise ValueError(f'service bundle kind must be "uns-service-bundle". Received {kind!r}.')

    metadata = _require_object(parsed.get("metadata"), "metadata")
    scaffold = _require_object(parsed.get("scaffold"), "scaffold")
    repository = _optional_object(parsed.get("repository"), "repository")
    domain = _optional_object(parsed.get("domain"), "domain")
    docs = _optional_object(parsed.get("docs"), "docs")
    _optional_object(parsed.get("provenance"), "provenance")

    bundle = ServiceBundle(
        schema_version=1,
        kind="uns-service-bundle",
        metadata=BundleMetadata(
            name=_require_non_empty_string(metadata.get("name"), "metadata.name"),
            display_name=_optional_non_empty_string(metadata.get("displayName"), "metadata.displayName"),
            service_type=_optional_non_empty_string(metadata.get("serviceType"), "metadata.serviceType"),
            summary=_optional_non_empty_string(metadata.get("summary"), "metadata.summary"),
            description=_optional_non_empty_string(metadata.get("description"), "metadata.description"),
            owner=_optional_non_empty_string(metadata.get("owner"), "metadata.owner"),
            tags=_optional_string_list(metadata.get("tags"), "metadata.tags"),
        ),
        scaffold=BundleScaffold(
            stack=_require_non_empty_string(scaffold.get("stack"), "scaffold.stack"),
            template=_require_non_empty_string(scaffold.get("template"), "scaffold.template"),
            features=_dedupe(_optional_string_list(scaffold.get("features"), "scaffold.features")),
        ),
        repository=BundleRepository(
            provider=_optional_non_empty_string(repository.get("provider"), "repository.provider") if repository else None,
            organization=_optional_non_empty_string(repository.get("organization"), "repository.organization") if repository else None,
            project=_optional_non_empty_string(repository.get("project"), "repository.project") if repository else None,
            repository=_optional_non_empty_string(repository.get("repository"), "repository.repository") if repository else None,
            default_branch=_optional_non_empty_string(repository.get("defaultBranch"), "repository.defaultBranch") if repository else None,
        ) if repository else None,
        domain=BundleDomain(
            inputs=_optional_list(domain.get("inputs"), "domain.inputs"),
            outputs=_optional_list(domain.get("outputs"), "domain.outputs"),
        ) if domain else None,
        docs=BundleDocs(
            service_spec=BundleServiceSpecDocs(
                goals=_optional_string_list(_optional_object(docs.get("serviceSpec"), "docs.serviceSpec").get("goals") if docs and _optional_object(docs.get("serviceSpec"), "docs.serviceSpec") else None, "docs.serviceSpec.goals"),
                non_goals=_optional_string_list(_optional_object(docs.get("serviceSpec"), "docs.serviceSpec").get("nonGoals") if docs and _optional_object(docs.get("serviceSpec"), "docs.serviceSpec") else None, "docs.serviceSpec.nonGoals"),
                acceptance_criteria=_optional_string_list(_optional_object(docs.get("serviceSpec"), "docs.serviceSpec").get("acceptanceCriteria") if docs and _optional_object(docs.get("serviceSpec"), "docs.serviceSpec") else None, "docs.serviceSpec.acceptanceCriteria"),
            ),
            agents=BundleAgentsDocs(
                project_context=_optional_string_list(_optional_object(docs.get("agents"), "docs.agents").get("projectContext") if docs and _optional_object(docs.get("agents"), "docs.agents") else None, "docs.agents.projectContext"),
                guardrails=_optional_string_list(_optional_object(docs.get("agents"), "docs.agents").get("guardrails") if docs and _optional_object(docs.get("agents"), "docs.agents") else None, "docs.agents.guardrails"),
                first_tasks=_optional_string_list(_optional_object(docs.get("agents"), "docs.agents").get("firstTasks") if docs and _optional_object(docs.get("agents"), "docs.agents") else None, "docs.agents.firstTasks"),
                verification=_optional_string_list(_optional_object(docs.get("agents"), "docs.agents").get("verification") if docs and _optional_object(docs.get("agents"), "docs.agents") else None, "docs.agents.verification"),
            ),
        ),
    )

    if bundle.scaffold.stack != expected_stack:
        raise ValueError(
            f'Bundle scaffold.stack is "{bundle.scaffold.stack}". Use {counterpart_cli_name} create --bundle <path> instead of {cli_name}.'
        )

    if bundle.scaffold.template != "default":
        raise ValueError(
            f'Bundle scaffold.template must be "default" for this MVP. Received "{bundle.scaffold.template}".'
        )

    return bundle


def _require_object(value: Any, path: str) -> Dict[str, Any]:
    if not isinstance(value, dict):
        raise ValueError(f"{path} must be an object.")
    return value


def _optional_object(value: Any, path: str) -> Optional[Dict[str, Any]]:
    if value is None:
        return None
    return _require_object(value, path)


def _require_non_empty_string(value: Any, path: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"{path} is required and must be a non-empty string.")
    return value.strip()


def _optional_non_empty_string(value: Any, path: str) -> Optional[str]:
    if value is None or value == "":
        return None
    return _require_non_empty_string(value, path)


def _optional_string_list(value: Any, path: str) -> List[str]:
    if value is None:
        return []
    if not isinstance(value, list):
        raise ValueError(f"{path} must be an array of strings.")
    return [_require_non_empty_string(item, f"{path}[{index}]") for index, item in enumerate(value)]


def _optional_list(value: Any, path: str) -> List[Any]:
    if value is None:
        return []
    if not isinstance(value, list):
        raise ValueError(f"{path} must be an array.")
    return value


def _dedupe(items: List[str]) -> List[str]:
    seen = set()
    output: List[str] = []
    for item in items:
        if item not in seen:
            seen.add(item)
            output.append(item)
    return output


def _render_key_value_list(entries: List[Tuple[str, Optional[str]]]) -> List[str]:
    return [f"- {label}: {value if value else 'Not specified'}" for label, value in entries]


def _render_string_list(items: List[str]) -> List[str]:
    if not items:
        return ["- None specified."]
    return [f"- {item}" for item in items]


def _render_unknown_list(items: List[Any]) -> List[str]:
    if not items:
        return ["- None specified."]

    rendered: List[str] = []
    for item in items:
        if isinstance(item, str) and item.strip():
            rendered.append(f"- {item}")
        else:
            rendered.append(f"- `{json.dumps(item, sort_keys=True)}`")
    return rendered

File: packages/uns-cli/src/service-bundle.ts
import { readFile } from "node:fs/promises";

export type SupportedBundleStack = "ts" | "python";

export type ServiceBundle = {
  schemaVersion: 1;
  kind: "uns-service-bundle";
  metadata: {
    name: string;
    displayName?: string;
    serviceType?: string;
    summary?: string;
    description?: string;
    owner?: string;
    tags: string[];
  };
  scaffold: {
    stack: SupportedBundleStack | string;
    template: string;
    features: string[];
  };
  repository?: {
    provider?: string;
    organization?: string;
    project?: string;
    repository?: string;
    defaultBranch?: string;
  };
  domain?: {
    inputs: unknown[];
    outputs: unknown[];
  };
  docs: {
    serviceSpec: {
      goals: string[];
      nonGoals: string[];
      acceptanceCriteria: string[];
    };
    agents: {
      projectContext: string[];
      guardrails: string[];
      firstTasks: string[];
      verification: string[];
    };
  };
};

type JsonRecord = Record<string, unknown>;

export type ServiceBundlePathParts = {
  topicPrefix: string;
  asset: string;
  objectType: string;
  objectId: string;
  attribute: string;
};

export type ServiceBundlePublisherContract = {
  mode: "uns-topic-publish";
  fullPath: string;
  pathParts: ServiceBundlePathParts;
  validityMode: "interval";
  expectedIntervalMs: number;
};

type ReadBundleOptions = {
  expectedStack: SupportedBundleStack;
  cliName: string;
  counterpartCliName: string;
};

const DEFAULT_PUBLISHER_EXPECTED_INTERVAL_MS = 60_000;

export async function readAndValidateServiceBundle(
  bundlePath: string,
  options: ReadBundleOptions,
): Promise<{ bundle: ServiceBundle; raw: string }> {
  const raw = await readFile(bundlePath, "utf8");
  const parsed = parseJsonObject(raw, bundlePath);
  const bundle = validateServiceBundle(parsed, options);
  return { bundle, raw };
}

export function generateServiceSpecMarkdown(bundle: ServiceBundle): string {
  const lines = [
    "# SERVICE_SPEC",
    "",
    "> Generated from `service.bundle.json`. Update the bundle as the source of truth.",
    "",
    "## Service Identity",
    ...renderKeyValueList([
      ["Name", bundle.metadata.name],
      ["Display Name", bundle.metadata.displayName],
      ["Service Type", bundle.metadata.serviceType],
      ["Owner", bundle.metadata.owner],
      ["Tags", bundle.metadata.tags.length ? bundle.metadata.tags.join(", ") : "None"],
    ]),
    "",
    "## Summary",
    "",
    bundle.metadata.summary?.trim() || "Not provided.",
    "",
    "## Description",
    "",
    bundle.metadata.description?.trim() || "Not provided.",
    "",
    "## Scaffold",
    ...renderKeyValueList([
      ["Stack", bundle.scaffold.stack],
      ["Template", bundle.scaffold.template],
      ["Features", bundle.scaffold.features.length ? bundle.scaffold.features.join(", ") : "None"],
    ]),
  ];

  if (bundle.repository) {
    lines.push(
      "",
      "## Repository",
      ...renderKeyValueList([
        ["Provider", bundle.repository.provider],
        ["Organization", bundle.repository.organization],
        ["Project", bundle.repository.project],
        ["Repository", bundle.repository.repository],
        ["Default Branch", bundle.repository.defaultBranch],
      ]),
    );
  }

  if (bundle.domain) {
    const outputPublisherContracts = getServiceBundleOutputPublisherContracts(bundle);
    lines.push(
      "",
      "## Domain Inputs",
      ...renderUnknownList(bundle.domain.inputs),
      "",
      "## Domain Outputs",
      ...renderUnknownList(bundle.domain.outputs),
    );
    if (outputPublisherContracts.length) {
      lines.push(
        "",
        "## Output Publisher Contracts",
        ...outputPublisherContracts.map(
          (contract) =>
            `- ${contract.fullPath}: validityMode: "interval", expectedIntervalMs: ${contract.expectedIntervalMs}`,
        ),
      );
    }
  }

  lines.push(
    "",
    "## Goals",
    ...renderStringList(bundle.docs.serviceSpec.goals),
    "",
    "## Non-Goals",
    ...renderStringList(bundle.docs.serviceSpec.nonGoals),
    "",
    "## Acceptance Criteria",
    ...renderStringList(bundle.docs.serviceSpec.acceptanceCriteria),
    "",
  );

  return lines.join("\n");
}

export function generateAgentsMarkdown(bundle: ServiceBundle): string {
  const outputPublisherContracts = getServiceBundleOutputPublisherContracts(bundle);
  const lines = [
    "# AGENTS",
    "",
    "> This repository was bootstrapped from `service.bundle.json`. Regenerate derived docs from the bundle instead of treating this file as the source of truth.",
    "",
    "## Service",
    ...renderKeyValueList([
      ["Name", bundle.metadata.name],
      ["Display Name", bundle.metadata.displayName],
      ["Stack", bundle.scaffold.stack],
      ["Template", bundle.scaffold.template],
    ]),
    "",
    "## Bundle Source",
    "",
    "- Read `service.bundle.json` first when planning or generating service-specific code; it is the project source of truth.",
    "- Keep `SERVICE_SPEC.md` and this file aligned with `service.bundle.json` when the bundle changes.",
    "",
  ];

  if (outputPublisherContracts.length) {
    lines.push(
      "## Output Publishing",
      ...outputPublisherContracts.flatMap((contract) => [
        `- ${contract.fullPath} publishes interval data with validityMode: "interval" and expectedIntervalMs: ${contract.expectedIntervalMs}.`,
        "- Do not publish interval attributes without both fields.",
      ]),
      "",
    );
  }

  lines.push(
    "## Project Context",
    ...renderStringList(bundle.docs.agents.projectContext),
    "",
    "## Guardrails",
    ...renderStringList(bundle.docs.agents.guardrails),
    "",
    "## First Tasks",
    ...renderStringList(bundle.docs.agents.firstTasks),
    "",
    "## Verification",
    ...renderStringList(bundle.docs.agents.verification),
    "",
  );

  return lines.join("\n");
}

export function getServiceBundleOutputPublisherContracts(bundle: ServiceBundle): ServiceBundlePublisherContract[] {
  const outputs = Array.isArray(bundle.domain?.outputs) ? bundle.domain.outputs : [];
  return outputs
    .map((output) => readOutputPublisherContract(output))
    .filter((contract): contract is ServiceBundlePublisherContract => Boolean(contract));
}

function readOutputPublisherContract(output: unknown): ServiceBundlePublisherContract | null {
  if (!isRecord(output)) {
    return null;
  }
  const rawPublisherContract = isRecord(output.publisherContract) ? output.publisherContract : null;
  if (!rawPublisherContract) {
    return null;
  }
  const mode = readOptionalString(rawPublisherContract.mode);
  if (mode && mode !== "uns-topic-publish") {
    return null;
  }
  const validityMode = readOptionalString(rawPublisherContract.validityMode);
  if (validityMode && validityMode !== "interval") {
    return null;
  }
  const fullPath = readOptionalString(rawPublisherContract.fullPath) ?? readOptionalString(output.topic);
  const pathParts = readPathParts(rawPublisherContract.pathParts) ?? readPathParts(output.pathParts);
  if (!fullPath || !pathParts) {
    return null;
  }
  return {
    mode: "uns-topic-publish",
    fullPath,
    pathParts,
    validityMode: "interval",
    expectedIntervalMs:
      readPositiveInteger(rawPublisherContract.expectedIntervalMs) ??
      readPositiveInteger(output.expectedIntervalMs) ??
      DEFAULT_PUBLISHER_EXPECTED_INTERVAL_MS,
  };
}

function readPathParts(value: unknown): ServiceBundlePathParts | null {
  if (!isRecord(value)) {
    return null;
  }
  const topicPrefix = readOptionalString(value.topicPrefix);
  const asset = readOptionalString(value.asset);
  const objectType = readOptionalString(value.objectType);
  const objectId = readOptionalString(value.objectId);
  const attribute = readOptionalString(value.attribute);
  if (!topicPrefix || !asset || !objectType || !objectId || !attribute) {
    return null;
  }
  return { topicPrefix, asset, objectType, objectId, attribute };
}

function readOptionalString(value: unknown): string | undefined {
  return typeof value === "string" && value.trim().length ? value.trim() : undefined;
}

function readPositiveInteger(value: unknown): number | undefined {
  const parsed =
    typeof value === "number"
      ? value
      : typeof value === "string" && value.trim().length
        ? Number(value.trim())
        : NaN;
  if (!Number.isFinite(parsed) || parsed <= 0) {
    return undefined;
  }
  return Math.floor(parsed);
}

function isRecord(value: unknown): value is JsonRecord {
  return Boolean(value && typeof value === "object" && !Array.isArray(value));
}

function validateServiceBundle(value: JsonRecord, options: ReadBundleOptions): ServiceBundle {
  const schemaVersion = value.schemaVersion;
  if (schemaVersion !== 1) {
    throw new Error(`service bundle schemaVersion must be 1. Received ${JSON.stringify(schemaVersion)}.`);
  }

  const kind = value.kind;
  if (kind !== "uns-service-bundle") {
    throw new Error(`service bundle kind must be "uns-service-bundle". Received ${JSON.stringify(kind)}.`);
  }

  const metadata = requireObject(value.metadata, "metadata");
  const scaffold = requireObject(value.scaffold, "scaffold");
  const repository = optionalObject(value.repository, "repository");
  const domain = optionalObject(value.domain, "domain");
  const docs = optionalObject(value.docs, "docs");
  optionalObject(value.provenance, "provenance");

  const normalized: ServiceBundle = {
    schemaVersion: 1,
    kind: "uns-service-bundle",
    metadata: {
      name: requireNonEmptyString(metadata.name, "metadata.name"),
      displayName: optionalNonEmptyString(metadata.displayName, "metadata.displayName"),
      serviceType: optionalNonEmptyString(metadata.serviceType, "metadata.serviceType"),
      summary: optionalNonEmptyString(metadata.summary, "metadata.summary"),
      description: optionalNonEmptyString(metadata.description, "metadata.description"),
      owner: optionalNonEmptyString(metadata.owner, "metadata.owner"),
      tags: optionalStringArray(metadata.tags, "metadata.tags"),
    },
    scaffold: {
      stack: requireNonEmptyString(scaffold.stack, "scaffold.stack"),
      template: requireNonEmptyString(scaffold.template, "scaffold.template"),
      features: dedupeStrings(optionalStringArray(scaffold.features, "scaffold.features")),
    },
    repository: repository
      ? {
          provider: optionalNonEmptyString(repository.provider, "repository.provider"),
          organization: optionalNonEmptyString(repository.organization, "repository.organization"),
          project: optionalNonEmptyString(repository.project, "repository.project"),
          repository: optionalNonEmptyString(repository.repository, "repository.repository"),
          defaultBranch: optionalNonEmptyString(repository.defaultBranch, "repository.defaultBranch"),
        }
      : undefined,
    domain: domain
      ? {
          inputs: optionalUnknownArray(domain.inputs, "domain.inputs"),
          outputs: optionalUnknownArray(domain.outputs, "domain.outputs"),
        }
      : undefined,
    docs: {
      serviceSpec: normalizeServiceSpecDocs(optionalObject(docs?.serviceSpec, "docs.serviceSpec")),
      agents: normalizeAgentsDocs(optionalObject(docs?.agents, "docs.agents")),
    },
  };

  if (normalized.scaffold.stack !== options.expectedStack) {
    throw new Error(
      `Bundle scaffold.stack is "${normalized.scaffold.stack}". Use ${options.counterpartCliName} create --bundle <path> instead of ${options.cliName}.`,
    );
  }

  if (normalized.scaffold.template !== "default") {
    throw new Error(
      `Bundle scaffold.template must be "default" for this MVP. Received "${normalized.scaffold.template}".`,
    );
  }

  return normalized;
}

function normalizeServiceSpecDocs(value?: JsonRecord): ServiceBundle["docs"]["serviceSpec"] {
  return {
    goals: optionalStringArray(value?.goals, "docs.serviceSpec.goals"),
    nonGoals: optionalStringArray(value?.nonGoals, "docs.serviceSpec.nonGoals"),
    acceptanceCriteria: optionalStringArray(value?.acceptanceCriteria, "docs.serviceSpec.acceptanceCriteria"),
  };
}

function normalizeAgentsDocs(value?: JsonRecord): ServiceBundle["docs"]["agents"] {
  return {
    projectContext: optionalStringArray(value?.projectContext, "docs.agents.projectContext"),
    guardrails: optionalStringArray(value?.guardrails, "docs.agents.guardrails"),
    firstTasks: optionalStringArray(value?.firstTasks, "docs.agents.firstTasks"),
    verification: optionalStringArray(value?.verification, "docs.agents.verification"),
  };
}

function parseJsonObject(raw: string, bundlePath: string): JsonRecord {
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (error) {
    throw new Error(`Failed to parse bundle JSON at ${bundlePath}: ${(error as Error).message}`);
  }

  return requireObject(parsed, "bundle");
}

function requireObject(value: unknown, path: string): JsonRecord {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`${path} must be an object.`);
  }
  return value as JsonRecord;
}

function optionalObject(value: unknown, path: string): JsonRecord | undefined {
  if (value === undefined || value === null) {
    return undefined;
  }
  return requireObject(value, path);
}

function requireNonEmptyString(value: unknown, path: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${path} is required and must be a non-empty string.`);
  }
  return value.trim();
}

function optionalNonEmptyString(value: unknown, path: string): string | undefined {
  if (value === undefined || value === null || value === "") {
    return undefined;
  }
  return requireNonEmptyString(value, path);
}

function optionalStringArray(value: unknown, path: string): string[] {
  if (value === undefined || value === null) {
    return [];
  }
  if (!Array.isArray(value)) {
    throw new Error(`${path} must be an array of strings.`);
  }
  return value.map((item, index) => requireNonEmptyString(item, `${path}[${index}]`));
}

function optionalUnknownArray(value: unknown, path: string): unknown[] {
  if (value === undefined || value === null) {
    return [];
  }
  if (!Array.isArray(value)) {
    throw new Error(`${path} must be an array.`);
  }
  return value;
}

function dedupeStrings(values: string[]): string[] {
  const seen = new Set<string>();
  const deduped: string[] = [];
  for (const value of values) {
    if (!seen.has(value)) {
      seen.add(value);
      deduped.push(value);
    }
  }
  return deduped;
}

function renderKeyValueList(entries: Array<[string, string | undefined]>): string[] {
  return entries.map(([label, value]) => `- ${label}: ${value && value.trim() ? value : "Not specified"}`);
}

function renderStringList(items: string[]): string[] {
  if (!items.length) {
    return ["- None specified."];
  }
  return items.map((item) => `- ${item}`);
}

function renderUnknownList(items: unknown[]): string[] {
  if (!items.length) {
    return ["- None specified."];
  }

  return items.map((item) => {
    if (typeof item === "string" && item.trim()) {
      return `- ${item}`;
    }
    return `- \`${JSON.stringify(item)}\``;
  });
}
